File: training/train on 1600-1800 problems/C_Minimizing_the_Sum.cpp
#include <bits/stdc++.h>
#define ll long long
using namespace std;

int n,k;
ll dp[300005][11];
vector<int> a;

ll solve(int indx,int curk){
    if (indx==n) return 0;
    if (dp[indx][curk]!=-1) return dp[indx][curk];
    ll res=0;
    int curmin=a[indx];
    ll cursum=0;
    for (int i=indx;i<min(n,indx+curk+1);i++) {
        curmin=min(curmin,a[i]);
        cursum+=a[i];
        res=max(res,cursum-((ll)(i-indx+1)*curmin)+solve(i+1,curk-(i-indx)));
    }
    return dp[indx][curk]=res;
}

int main(){
    ios::sync_with_stdio(0);
    cin.tie(0);
    int t;
    cin>>t;
    while (t--){
        cin>>n>>k;
        a.resize(n);
        for (int i=0;i<=n;i++){
            for (int j=0;j<=k;j++){
                dp[i][j]=-1;
            }
        }
        for(int i=0;i<n;i++){
            cin>>a[i];
            }
            cout<<accumulate(a.begin(),a.end(),0LL)-solve(0,k)<<"\n";
    }
    return 0;
}

File: training/train on 1600-1800 problems/C_Minimizing_the_Sum.py
import sys
sys.setrecursionlimit(300000)

def solve(indx,curk):
    if indx==n:
        return 0
    if dp[indx][curk]!=-1:
        return dp[indx][curk]
    res=0
    curmin=a[indx]
    cursum=0
    for i in range(indx,min(n,indx+curk+1)):
        curmin=min(curmin,a[i])
        cursum+=a[i]
        res=max(res,cursum-((i-indx+1)*curmin)+solve(i+1,curk-(i-indx)))
    dp[indx][curk]=res
    return res

t=int(input())
for _ in range(t):
    n,k=map(int,input().split())
    a=list(map(int,input().split()))
    dp=[[-1]*(k+1) for _ in range(n+1)]
    print(sum(a)-solve(0,k))